controller/src/main/java/io/pravega/controller/server/rpc/auth/StrongPasswordProcessor.java

```diff
@@ -0,0 +1,101 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.controller.server.rpc.auth;
+
+import java.math.BigInteger;
+import java.security.NoSuchAlgorithmException;
+import java.security.SecureRandom;
+import java.security.spec.InvalidKeySpecException;
+import javax.crypto.SecretKeyFactory;
+import javax.crypto.spec.PBEKeySpec;
+
+/**
+ * This class implements a a `PBKDF2WithHmacSHA1` based password digest creator and validator.
+ *
+ * Following steps are taken during the creation of the digest:
+ *  1. A salt is generated.
+ *  2. The password is encoded with this salt.
+ *  3. Number of iterations, salt and this password is string encoded and concatenated with ":" as separator.
+ *  4. This whole string is again string encoded with base 16.
+ *
+ *  For validation these steps are reversed to get the password digest from the stored password. The incoming password
+ *  is digested with the retrieved iterations and salt. The generated digest is then cross checked against the created digest.
+ */
+public class StrongPasswordProcessor {
+
+    private final int iterations = 1000;
+
+    /*
+     * @param password              The incoming password.
+     * @param encryptedPassword     The stored password digest.
+     * @return                      true if the password matches, false otherwise.
+     * @throws NoSuchAlgorithmException encryption exceptions.
+     * @throws InvalidKeySpecException encryption exceptions.
+     */
+    public boolean checkPassword(String password, String encryptedPassword) throws NoSuchAlgorithmException, InvalidKeySpecException {
+        String storedPassword = new String(fromHex(encryptedPassword));
+        String[] parts = storedPassword.split(":");
+        int iterations = Integer.parseInt(parts[0]);
+        byte[] salt = fromHex(parts[1]);
+        byte[] hash = fromHex(parts[2]);
+
+        PBEKeySpec spec = new PBEKeySpec(password.toCharArray(), salt, iterations, hash.length * 8);
+        SecretKeyFactory skf = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA1");
+        byte[] testHash = skf.generateSecret(spec).getEncoded();
+
+        int diff = hash.length ^ testHash.length;
+        for (int i = 0; i < hash.length && i < testHash.length; i++) {
+            diff |= hash[i] ^ testHash[i];
+        }
+        return diff == 0;
+    }
+
+    /*
+     * @param userPassword The password to be encrypted.
+     * @return The encrypted string digest.
+     * @throws NoSuchAlgorithmException encryption exceptions.
+     * @throws InvalidKeySpecException encryption exceptions.
+     */
+    public String encryptPassword(String userPassword) throws NoSuchAlgorithmException, InvalidKeySpecException {
+        char[] chars = userPassword.toCharArray();
+        byte[] salt = getSalt();
+
+        PBEKeySpec spec = new PBEKeySpec(chars, salt, iterations, 64 * 8);
```

Comment:
Can you make the iterations, size, algorithm and salt size, variables. Then we could create a unit test that does a sanity check by verifying the hash changes if you change any of them. This would make sure the parameters aren't ignored. (I found in testing with some algorithms they are)

Reply:
Fixed. Will add the said test as part of another PR.